### ISSUE TITLE
GRPO Trainer + Peft + Gradient checkpointing doesn't work

### ISSUE THREAD

**Issue by kimihailvfg** [huggingface/trl#3089]
### Reproduction

```python

from trl import GRPOConfig, GRPOTrainer, ModelConfig, get_peft_config

training_args = GRPOConfig(
            learning_rate=5e-6,
            adam_beta1=0.9,
            adam_beta2=0.99,
            weight_decay=0.01,
            warmup_ratio=0.1,
            lr_scheduler_type="cosine",
            optim="adamw_torch",
            logging_steps=1,
            per_device_train_batch_size=16,
            gradient_accumulation_steps=1
            num_generations=8,
            max_prompt_length=256,
            max_completion_length=512,
            num_train_epochs=1,
            save_steps=100,
            max_grad_norm=0.1,
            report_to="wandb",
            output_dir="out",
            log_completions=True,
            run_name="grpo",
            use_vllm=True,
            vllm_gpu_memory_utilization=0.6,
            gradient_checkpointing=True
        )

model_config = ModelConfig(
    model_name_or_path="meta-llama/Llama-3.2-3B-Instruct",
    torch_dtype="bfloat16",
    attn_implementation="flash_attention_2",
    use_peft=True,
    load_in_4bit=False,
    lora_r=64,
    lora_target_modules=[
        "q_proj",
        "k_proj",
        "v_proj",
        "o_proj",
        "gate_proj",
        "up_proj",
        "down_proj",
    ],
    lora_alpha=1,
    use_rslora=True
)

trainer = GRPOTrainer(
    model=model_config.model_name_or_path,
    reward_funcs=[reward_funcs.format_reward],
    args=training_args,
    train_dataset=dataset,
    peft_config=get_peft_config(model_config)
)```

### System Info

- Platform: Linux-5.15.0-126-generic-x86_64-with-glibc2.35
- Python version: 3.11.11
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3
- Transformers version: 4.49.0
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.2.0
- HF Hub version: 0.27.0
- TRL version: 0.15.2
- bitsandbytes version: 0.45.3
- DeepSpeed version: not installed
- Diffusers version: 0.32.2
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.61.1
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by tchang1997** [comment]
Could you define "doesn't work?" I have training running with PeFT + gradient checkpointing without issues but had to play around with the settings.

At a glance, the only major difference I see between our configs is that you might need `gradient_checkpointing_kwargs=dict(reentrant=True)` in your `GRPOConfig`.

**Comment by binary-husky** [comment]
you must set `reentrant=True`

**Comment by leosmith8004** [comment]
OMG，i have the same issue with you，have you solved it ？ Thanks for your reply

**Comment by asadrufat** [comment]
I have the same issue. Any updates?

**Comment by shimudong** [comment]
> I have the same issue. Any updates?

I tried to set use_reentrant=True, but it seems doesn't work for me. 
I just put in a quick hard-coded fix inside the training_step function in transformers/trainer.py. Basically, I added this snippet:
```python
for k,v in inputs.items():
    if type(v) == torch.Tensor:
        if v.dtype==torch.float32:
            inputs[k].requires_grad=True
``` 
The code works for me. 

I think you can use monkey-patch to insert the code.

I guess the bug is caused by the difference of rollout and logp calculation.

**Comment by WooperWise** [comment]
Double check your GPU supports bf16, if it doesn't, load your model in f32. I have just tested this and it helped me.

**Comment by behroozazarkhalili** [comment]
This looks resolved on current `main`. I dug into the **why** and verified it with a minimal reproduction.

## Why it failed (TRL 0.15.2)

With a PEFT/LoRA adapter, the base model is frozen — only the adapter weights require grad, so the input embeddings have `requires_grad=False`. **Reentrant** gradient checkpointing (`use_reentrant=True`, which you set) recomputes the forward pass during backward and needs at least one grad-requiring input to rebuild the autograd graph. With frozen embeddings and no input-gradient hook, the recomputed activations have no `grad_fn`, hence `element 0 of tensors does not require grad and does not have a grad_fn`. It worked without PEFT because there the base weights are trainable.

At v0.15.2 `GRPOTrainer` did not call `model.enable_input_require_grads()` (the standard remedy), so the reentrant + PEFT path failed.

## Why it works now

Current `main` is protected by **two independent mechanisms, either of which is sufficient**:

1. `grpo_trainer.py` enables input grads for the exact case:
   ```python
   if is_peft_model(model) and args.gradient_checkpointing:
       model.enable_input_require_grads()
   ```
   (added in #4577 / #4723; the comment notes `transformers.Trainer` normally does this but is currently affected by huggingface/transformers#42489, so TRL does it itself.)
2. `use_reentrant` now defaults to `False` (set in #4951), and the non-reentrant path doesn't need a grad-requiring input at all.

## Verified by reproduction

Minimal GRPO + LoRA + `gradient_checkpointing=True` run (tiny model, CPU), as a 2×2 grid:

| | input-grad guard ON | guard OFF |
|---|---|---|
| `use_reentrant=True` (your setting) | ✅ trains | ❌ reproduces `element 0 … does not require grad` |
| `use_reentrant=False` (current default) | ✅ trains | ✅ trains |

The error reproduces **only** in the `use_reentrant=True` + no-guard cell — i.e. exactly the v0.15.2 situation. On current `main` your configuration trains fine, because the guard covers the reentrant path even if you explicitly set `use_reentrant=True`.

## Suggestion

I think this can be closed as resolved on `main`. For anyone still on an older release hitting this, the fixes are to either upgrade, or call `model.enable_input_require_grads()` on the PEFT model before training.
